src/freenet/pluginmanager/FredPluginFCPMessageHandler.java

```diff
@@ -0,0 +1,580 @@
+         * 
+         * You <b>must not</b> use this for constructing outgoing messages in server or client
+         * implementations.<br>
+         * 
+         * This function is typically to construct incoming messages for passing them to the message
+         * handling function
+         * {@link FredPluginFCPMessageHandler#handlePluginFCPMessage(FCPPluginClient,
+         * FCPPluginMessage)}.<br><br>
+         * 
+         * See the JavaDoc of the member variables with the same name as the parameters for an
+         * explanation of the parameters.<br>
+         */
+        public static FCPPluginMessage constructRawMessage(ClientPermissions permissions,
+                String identifier, SimpleFieldSet params, Bucket data, Boolean success,
+                String errorCode, String errorMessage) {
+            
+            return new FCPPluginMessage(permissions, identifier, params, data, success,
+                errorCode, errorMessage);
+        }
+
+        @Override
+        public String toString() {
+            return super.toString() +
+                " (permissions: " + permissions +
+                "; identifier: " + identifier +
+                "; data: " + data +
+                "; success: " + success +
+                "; errorCode: " + errorCode +
+                "; errorMessage: " + errorMessage + 
+                // At the end because a SimpleFieldSet usually contains multiple line breaks.
+                "; params: " + '\n' + (params != null ? params.toOrderedString() : null);
+        }
+    }
+
+
+    /**
+     * Message handling function for messages received from a plugin FCP server or client.<br/><br/>
+     * 
+     * <b>ATTENTION</b>: Please read the different constraints for server and client side message
+     * handlers at the child interfaces:<br/>
+     * - {@link ServerSideFCPMessageHandler#handlePluginFCPMessage(FCPPluginClient,
+     *   FCPPluginMessage)}<br/>
+     * - {@link ClientSideFCPMessageHandler#handlePluginFCPMessage(FCPPluginClient,
+     *   FCPPluginMessage)}<br/>
+     * 
+     * To stress those different constraints, you should also not implement this interface but one
+     * of the child interfaces {@link ServerSideFCPMessageHandler} and
+     * {@link ClientSideFCPMessageHandler}.
+     */
+    FCPPluginMessage handlePluginFCPMessage(FCPPluginClient client, FCPPluginMessage message);
+
+
+    /**
+     * Plugins which provide FCP services to clients must implement this interface.<br/>
+     * The purpose of this interface is to provide a message handling function for processing
+     * messages received from the clients.
+     * 
+     * @see FredPluginFCPMessageHandler
+     *          The parent interface FredPluginFCPMessageHandler provides an overview.
+     * @see ClientSideFCPMessageHandler
+     *          The opposite version of this interface for client plugins
+     */
+    public interface ServerSideFCPMessageHandler extends FredPluginFCPMessageHandler {
```

Comment:
Commit 8b1e652d8451b5817ee7e1d95893e5dad3f60420 moves FCPPluginMessage to its own file, which reduces the file size as suggested by the previous comment.

Also, addendum to my original reply:
You said:
> It needlessly complicates the interface name in external code.

Java allows import of member classes, so the external code can reference it by "ServerSideFCPMessageHandler" instead of "FredPluginFCPMessageHandler.ServerSideFCPMessageHandler".
The more lengthy version is used in some places of this pull request for making the code more understandable in its specific context. Plugin authors are IMHO fine to use the short version.